cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -1066,20 +1121,114 @@ class ObjectAppendStream final : public io::OutputStream {
       // flush. This also avoids some unhandled errors when flushing in the destructor.
       return Status::OK();
     }
-    return CommitBlockList(block_blob_client_, block_ids_, commit_block_list_options_);
+
+    Future<> pending_blocks_completed;
+    {
+      std::unique_lock<std::mutex> lock(upload_state_->mutex);
+      pending_blocks_completed = upload_state_->pending_blocks_completed;
+    }
+
+    RETURN_NOT_OK(pending_blocks_completed.status());
```

Comment:
Because we can not hold the lock while waiting for the future. Another instance will acquire the lock and mark the future as completed while we wait on the future.

Reply:
That makes sense indeed. Thank you!